File: tests/cs/dict.cs
using System;
using System.Diagnostics;
using System.Timers;
using System.Collections.Generic;

class HelloWorld {

    delegate void MyBlock();

    static void profile ( int count, string category, MyBlock f ) {
        long n = 1;
        double total = 0;
        while (true) {
            Stopwatch stopwatch = Stopwatch.StartNew();
            for ( long i = 0; i < n; i++ ) {
                f();
            }
            stopwatch.Stop();
            total = stopwatch.Elapsed.TotalSeconds;
            if (total >= 0.5 || n >= 1000000000L) break;
            double per = Math.Max(total / n, 1e-9);
            long next = (long)(0.5 / per * 1.2);
            next = Math.Min(next, 100L * n);
            next = Math.Max(next, n + 1);
            n = Math.Min(next, 1000000000L);
        }
        Console.WriteLine($"\"{category}\", {total / n}, {n}");
    }

    public static void MakeRandomSequence(ref string[] src)
    {
        int n = 500000;
        uint mod = (uint)n;
        src = new string[n]; // Resizing the array in C#
        for (int i = 0; i < n; i++)
        {
            uint num = (271828183u ^ (uint)(i * 119)) % mod;
            src[i] = num.ToString();
        }
    }

    public static int Dict(string[] src)
    {
        var tab = new Dictionary<string, int>();
        int maxOcc = 1;
        foreach (var s in src)
        {
            if ( tab.TryGetValue(s, out int val) )
            {
                val ++;
                maxOcc = Math.Max(val, maxOcc);
            }
            else
            {
                tab[s] = 1;
            }

        }
        return maxOcc;
    }

    static void Main() {
        int occ = 0;
        profile(10, "dictionary", () => {
            string[] sequence = null;
            MakeRandomSequence(ref sequence);
            int maxOccurrence = Dict(sequence);
            occ += maxOccurrence;
        });
        Debug.Assert(occ>13);
    }
}

File: tests/js/dict.js
// dictionary test

function dict_makeSrc(){
	var src = [];
	var n = 500000;
	var modn = n;
	for (var i=0; i != n; ++i ) {
		var num = (271828183 ^ i*119) % modn;
		src.push('_' + num);
	}
	return src;
}

function dict(src) {
	var tab = {}
	var n = src.length;
	var max = 1
	for (var i=0; i != n; ++i ) {
		var l = src[n];
		if ( tab.hasOwnProperty(l) ) {
			max = Math.max(++tab[l],max);
		} else {
			tab[l] = 1;
		}
	}
	return max;
}

// infrastructure

function timeStamp() {
	return performance.now();
}

function profile(tname, cnt, testFn) {
    let n = 1
    let total = 0
    while (true) {
        const start = timeStamp()
        for (let i = 0; i < n; ++i) {
            testFn()
        }
        total = timeStamp() - start
        if (total >= 500.0 || n >= 1000000000) break
        const per = Math.max(total / n, 1e-6)
        let next = Math.floor(500.0 / per * 1.2)
        next = Math.min(next, 100 * n)
        next = Math.max(next, n + 1)
        n = Math.min(next, 1000000000)
    }
    print('"' + tname + '", ' + (total / 1000.0 / n) + ', ' + n)
}

function performance_tests() {
	{
		var src = dict_makeSrc();
		profile("dictionary",10,function(){
			dict(src);
		});
	}
	timeStamp();
}

performance_tests();
